ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockValidator.java

```diff
@@ -79,6 +81,25 @@ public BlockValidator(final Spec spec, RecentChainData recentChainData) {
       return completedFuture(reject("Block does not descend from finalized checkpoint"));
     }
 
+    Optional<ExecutionPayload> maybeExecutionPayload =
+        block.getMessage().getBody().getOptionalExecutionPayload();
+    Optional<BeaconState> maybeState = recentChainData.getBestState();
```

Comment:
We can't use `recentChainData.getBestState()` here because this block might not be on the same fork as our best state.  Further down (line 120) or so we get a state based on the parent block which is what we need to use to check if the merge has been completed and a non-default payload is required.

Reply:
Given we need a state to verify the timestamp anyway, I'd just push the whole timestamp check down to where we have the relevant state and we can then implement it exactly as the spec does - if mergeComplete(state) then require timestamp == computeTimeAtSlot(..).  No need to explicitly check if the execution payload is default or not because if it is then the timestamp won't match.

Reply:
> We can't use recentChainData.getBestState() here because this block might not be on the same fork as our best state

i was thinking that it was acceptable since under the hood it is only getting the genesis time

> No need to explicitly check if the execution payload is default or not because if it is then the timestamp won't match.

right, also because i suspect isDefault isn't cheep as it may seems

Reply:
True, we would get away with that but we've had a lot of problems in the past because we wound up using the wrong state for things.   We could though just use `recentChainData.getGenesisTime` and have a variant of `computeTimeAtSlot` which takes the genesis time instead of the state (which is often useful anyway). Then we're being explicit about what we're actually using and there won't be any confusion.  We'd still need to check if the payload is non-default if the pre-state had completed the merge though so it's probably more efficient to just do the one check once we have the state.

Reply:
And yes, that is different to what I was thinking when we were first talking about this. Sorry.